packages/backend/src/services/ProjectService/ProjectService.ts

```diff
@@ -122,32 +130,60 @@ export class ProjectService {
         this.jobModel = jobModel;
         this.emailClient = emailClient;
         this.spaceModel = spaceModel;
+        this.sshKeyPairModel = sshKeyPairModel;
     }
 
-    private async _getWarehouseClient(
-        projectUuid: string,
-    ): Promise<WarehouseClient> {
+    private async _resolveWarehouseClientSshKeys<
+        T extends { warehouseConnection: CreateWarehouseCredentials },
+    >(args: T): Promise<T> {
+        if (
+            (args.warehouseConnection.type === WarehouseTypes.REDSHIFT ||
+                args.warehouseConnection.type === WarehouseTypes.POSTGRES) &&
+            args.warehouseConnection.useSshTunnel
+        ) {
+            const publicKey = args.warehouseConnection.sshTunnelPublicKey || '';
+            const { privateKey } = await this.sshKeyPairModel.get(publicKey);
+            return {
+                ...args,
+                warehouseConnection: {
+                    ...args.warehouseConnection,
+                    sshTunnelPrivateKey: privateKey,
+                },
+            };
+        }
+        return args;
+    }
+
+    private async _getWarehouseClient(projectUuid: string): Promise<{
```

Comment:
perhaps this function should become a wrapper, so the tunnel closes once the method finishes, or return a callback to close the client